contrib/vcloud/benchmarkclient_executor.py

```diff
@@ -35,12 +37,95 @@ def set_vcloud_jar_path(p):
+        self.container_mount_point = container_mount_point
+
+    def find_executable(self, executable_name, subdir=""):
+        logging.debug(
+            "Using custom tool locator to find executable %s", executable_name
+        )
+        assert (
+            os.path.basename(executable_name) == executable_name
+        ), "Executable needs to be a simple file name"
+        dirs = []
+
+        if not self.tool_directory:
+            raise ToolNotFoundException(
+                "Podman containerized tool info module execution is only possible with --tool-directory explicitly set."
+            )
+
+        assert self.container_mount_point is not None, "Container mount point not set"
+
+        # At this point we know, that the tool is located at container_mount_point
+        # as the container as the tool_dir mounted to this location
+        dirs.append(os.path.join(self.container_mount_point, subdir))
+        logging.debug("Searching for executable %s in %s", executable_name, dirs)
+
+        executable = benchexec.util.find_executable2(executable_name, dirs)
+        if executable:
+            return executable
+
+        other_file = benchexec.util.find_executable2(executable_name, dirs, os.F_OK)
+        if other_file:
+            raise ToolNotFoundException(
+                f"Could not find executable '{executable_name}', "
+                f"but found file '{other_file}' that is not executable."
+            )
+
+        msg = (
+            f"Could not find executable '{executable_name}'. "
+            f"The searched directories were: " + "".join("\n  " + d for d in dirs)
+        )
+        if not self.tool_directory:
+            msg += "\nYou can specify the tool's directory with --tool-directory."
+
+        raise ToolNotFoundException(msg)
+
+
 def init(config, benchmark):
     global _JustReprocessResults
     _JustReprocessResults = config.reprocessResults
-    tool_locator = benchexec.tooladapter.create_tool_locator(config)
-    benchmark.executable = benchmark.tool.executable(tool_locator)
-    benchmark.tool_version = benchmark.tool.version(benchmark.executable)
+
+    if config.containerImage:
+        from vcloud.podman_containerized_tool import TOOL_DIRECTORY_MOUNT_POINT
+
+        tool_locator = CustomToolLocator(
+            config.tool_directory, TOOL_DIRECTORY_MOUNT_POINT
+        )
+        executable_for_version = benchmark.tool.executable(tool_locator)
+        benchmark.tool_version = benchmark.tool.version(executable_for_version)
+
+        # If the tool info does not call find_executable, we don't know if the
+        # executable path is containing the mount point.
+        # In this case we can check whether the path is relative
+        # and continue with the assumption that it is relative to the provided
+        # tool directory.
+        try:
+            executable_relative_to_mount_point = Path(
+                executable_for_version
+            ).relative_to(TOOL_DIRECTORY_MOUNT_POINT)
```

Comment:
Here is what I considered:

1. executable_for_version is created with the find_tool helper

Then the path will be `TOOL_DIRECTORY_MOUNT_POINT/subdir/executable` and the
relative method will retrieve the `subdir/executable` part. We then know `tool_dir/subdir/executable` is the path to the executable on the host.

2. find tool helper is not called

a. The ti just returns a relative path, eg., ./sub/executable 
The relative_to will fail with ValueError; since ./sub/executable is not absolute,
it is assumed to be relative to the tool directory and we proceed like in 1.

b. The ti returns an absolute path, that is not calculated in the container (e.g. /opt/my/tool).
The relative_to() call with fail with ValueError. There is no hope of recovering from here.

c. There could be the an edge case, where someone hardcodes "/mnt/__benchexec_tool_dir/something" . If that something is invoking '..' to a point further up the path than "/mnt/__benchexec_tool_dir" will cause a ValueError again.

d. If the tool returns `../../something` this is currently not handled. If the executable indeed exists it will probably be uploaded to the BenchCloud.

If d. is a concern to us, then we could sanitize the path coming from the ti or be a lot stricter in what we acceptas valid return from tool.executable()

Reply:
> Here is what I considered:
> 
> 1. executable_for_version is created with the find_tool helper

What is the "find_tool helper"?

> Then the path will be `TOOL_DIRECTORY_MOUNT_POINT/subdir/executable` and the relative method will retrieve the `subdir/executable` part. We then know `tool_dir/subdir/executable` is the path to the executable on the host.

What is "the relative method"?

> 2. find tool helper is not called
> 
> a. The ti just returns a relative path, eg., ./sub/executable The relative_to will fail with ValueError; since ./sub/executable is not absolute, it is assumed to be relative to the tool directory and we proceed like in 1.
>
> b. The ti returns an absolute path, that is not calculated in the container (e.g. /opt/my/tool). The relative_to() call with fail with ValueError. There is no hope of recovering from here.

So we want to always return relative paths as they are, and for absolute paths call `relative_to`?
Wouldn't the following convey the intent more clearly?
```
executable_for_version = Path(executable_for_version)
# ensure executable_for_version is relative
if executable_for_version.is_absolute():
    try:
        executable_for_version = executable_for_version.relative_to(TOOL_DIRECTORY_MOUNT_POINT)
    except ValueError:
        raise ...
```

> d. If the tool returns `../../something` this is currently not handled. If the executable indeed exists it will probably be uploaded to the BenchCloud.
> 
> If d. is a concern to us, then we could sanitize the path coming from the ti or be a lot stricter in what we acceptas valid return from tool.executable()

If `executable_for_version` points to somewhere outside of the tool directory, this means it comes from the container. I think that causes trouble no matter whether the path exists outside of the container or not, because we could for example upload a file that is not the one we want to benchmark. And it is trivial to detect this case and reject it, so this should be done.

Reply:
Hi,
I changed the code to the one you suggested. I also added a check for d.